### PR TITLE
Use ch-go for low level implementation

### PR DESCRIPTION
**Pull request by gingerwizard**
Closes https://github.com/ClickHouse/clickhouse-go/issues/623

### CODE REVIEW

**Comment by gingerwizard** [comment]
Work in progress. Current proposed approach:

1. proto.Reader replaces the Decoder
2. proto.Buffer replaces the Encoder - this means we have to buffer and then send via Stream

Note we still use the Stream for now - eventually id like to remove this as well and switch in ch-go compression. Not the change in the Column interface `Interface`.

File: conn_handshake.go [hunk 1]
```diff
@@ -25,39 +25,32 @@ import (
...
+	defer c.buffer.Reset()
 	c.debugf("[handshake] -> %s", proto.ClientHandshake{})
 	c.conn.SetDeadline(time.Now().Add(c.opt.DialTimeout))
 	defer c.conn.SetDeadline(time.Time{})
 	{
-		c.encoder.Byte(proto.ClientHello)
-		if err := (&proto.ClientHandshake{}).Encode(c.encoder); err != nil {
-			return err
-		}
+		c.buffer.PutByte(proto.ClientHello)
```
Comment (ernado) [inline conn_handshake.go:33]: you can use chproto.ClientHello here

File: conn_http_query.go [hunk 2]
```diff
@@ -62,9 +63,10 @@ func (h *httpConnect) query(ctx context.Context, release func(*connect, error),
 
 	go func() {
 		for {
-			block, err := h.readData()
+			block, err := h.readData(reader)
 			if err != nil {
-				if err != io.EOF {
+				// ch-go wraps EOF files
```
Comment (ernado) [inline conn_http_query.go:68]: probably "ch-go wraps EOF errors"

**Review by ernado — APPROVED** [review]
Great job, looks good in general.

It's hard to fully review it tho.

I see that we still have proto package, hope that we will use ch-go/proto for that later.

File: conn_handshake.go (hunk 1 above)
Comment (gingerwizard) [reply conn_handshake.go:33]: Better to just rip out handshake code completely on next PR

**Pull request merged**